### DISCUSSION TITLE
mqtt_acl with redis: combining username & ipaddr possible?

### DISCUSSION THREAD

**Discussion by bogorad** [emqx/emqx#5986] [Q&A]
I use Redis for auth/ACL.

In my setup, a controller _can_ be programmed to provide a username/password combo, but a camera can not. Given that both the controller and the camera(s) are in the same location, I was thinking of doing this:

1. Authenticate the controller via username/password (ACL for the controller is pre-set)
2. Get the controller's IP address via a webhook
3. Allow some camera-specific topics from the controller's IP address

However, it looks like when using Redis for ACL, I can either use username or ipaddress - but not both.

Am I correct?

Any ideas on how to mitigate this would be much appreciated!

**Comment by k32** [comment]
Hello,

I think Postgres and MySQL ACL plugins allow to supply custom SQL queries, so maybe they will be flexible enough to support this usecase.
But from security perspective, if the controller and the camera are behind a NAT, this will open the topic for anyone who is behind the NAT. Also, this still vulnerable to IP spoofing.

**Reply by bogorad** [reply]
> But from security perspective, if the controller 
> and the camera are behind a NAT, this will open 
> the topic for anyone who is behind the NAT.

Well, the system knows which cams are at which location, so that is not a problem. Cams could monitor each other's topics, also no big deal.

> Also, this still vulnerable to IP spoofing.

Again, the attacker needs to know both the current IP address and cam's MAC address. Not likely.

But thank you for your thoughts.
